src/ucm/bistro/bistro.c

```diff
@@ -45,6 +48,64 @@ static ucs_status_t ucm_bistro_protect(void *addr, size_t len, int prot)
     return UCS_OK;
 }
 
+static double ucm_bistro_get_time_msec()
+{
+    struct timeval tv;
+
+    gettimeofday(&tv, NULL);
+    return (double)tv.tv_sec * UCS_MSEC_PER_SEC +
+           (tv.tv_usec / (UCS_USEC_PER_SEC / UCS_MSEC_PER_SEC));
+}
+
+void ucm_bistro_modify_code(void *dst, const ucm_bistro_lock_t *bytes)
+{
+    uint32_t value;
+
+    if (sizeof(value) > sizeof(*bytes)) {
+        memcpy(&value, dst, sizeof(value));
+    }
+
+    UCS_STATIC_ASSERT(sizeof(*bytes) <= sizeof(value));
+    memcpy(&value, bytes, sizeof(*bytes));
+    (void)ucs_atomic_swap32(dst, value);
+}
+
+ucs_status_t
+ucm_bistro_apply_patch_atomic(void *dst, const void *patch, size_t len)
+{
+    size_t skip           = sizeof(ucm_bistro_lock_t);
+    double grace_duration = 5.0;
+    double deadline;
+    ucs_status_t status;
+
+    status = ucm_bistro_protect(dst, len, UCM_PROT_READ_WRITE_EXEC);
+    if (UCS_STATUS_IS_ERR(status)) {
+        return status;
+    }
+
+    /* Lock the codepatch and wait for existing flows to complete */
+    ucm_bistro_patch_lock(dst);
+    ucs_clear_cache(dst, UCS_PTR_BYTE_OFFSET(dst, len));
+
+    deadline = ucm_bistro_get_time_msec() + grace_duration;
```

Comment:
IMO to void confusion, all time values in bistro and in the test should in "seconds" units and not msec

Reply:
fixed in test_hooks and bistro.c